webapp/handlers.py

```diff
@@ -0,0 +1,93 @@
+import socket
+from datetime import datetime
+from urllib.parse import unquote, urlparse, urlunparse
+
+import flask
+import webapp.template_utils as template_utils
+
+def set_handlers(app):
+    @app.context_processor
+    def utility_processor():
+        """
+        This defines the set of properties and functions that will be added
+        to the default context for processing templates. All these items
+        can be used in all templates
+        """
+
+        return {
+            # Functions
+            "format_date": template_utils.format_date,
+            "get_year": datetime.now,
+            "replace_admin": template_utils.replace_admin,
+            "versioned_static": template_utils.versioned_static,
+            "truncate_chars": template_utils.truncate_chars,
+        }
+
+    # Error handlers
+    # ===
+    @app.errorhandler(404)
+    def page_not_found(error):
+        """
+        For 404 pages, display the 404.html template,
+        passing through the error description.
+        """
+
+        return flask.render_template("404.html", error=error.description), 404
+
+    @app.errorhandler(500)
+    @app.errorhandler(501)
+    @app.errorhandler(502)
+    @app.errorhandler(503)
+    @app.errorhandler(504)
+    @app.errorhandler(505)
+    def internal_error(error):
+        error_name = getattr(error, "name", type(error).__name__)
+        status_code = getattr(error, "code", 500)
+
+        return (
+            flask.render_template("500.html", error_name=error_name),
+            status_code,
+        )
+
+    # Global tasks for all requests
+    # ===
+    @app.before_request
+    def clear_trailing():
```

Comment:
You probably don't want this function. URLs in the GUI appear to in fact have trailing slashes.

Reply:
The GUI handles all of the URL parsing but I think we might like the slashes removed pre-gui.

Reply:
Well, the GUI handles the URL parsing after the first load. But if you reload, then it's back to Flask, initially. So if you GUI crafts a URL like https://example.com/somewhere/, which works initially, but then you reload, then Flask redirects to remove that slash before allow JS to take over again, it could easily confuse the GUI.

If you're doing URL routing in the GUI SPA, I would imagine you want Flask messing with your URLs as little as possible.

Reply:
https://jujucharms.com/q/store/?type=charm is an example of a URL with a slash on the end

Reply:
Yeah good call. I'll standardize this between the two parsers.